src/graphnet/training/weight_fitting.py

```diff
@@ -0,0 +1,93 @@
+
+        Args:
+            bins (numpy.array): An array containing the bins. E.g. numpy.arange(0,180,1)
+            variable (str): the name of the truth variable in truth_table
+            weight_name (str, optional): the assigned column name of the weights if add_to_database = True. Defaults to None.
+            add_to_database (bool, optional): If True, the weights are written to the database as a new table named weight_name . Defaults to False.
+            selection (List, optional): a List containing the event_no's to which the weights are calculated. Defaults to None.
+
+        Returns:
+            _type_: pandas.DataFrame
+        """
+        truth = self._get_truth(variable, selection)
+        # Histogram `truth_values`
+        bin_counts, _ = np.histogram(truth[variable], bins=bins)
+
+        # Get reweighting for each bin to achieve uniformity. (NB: No normalisation applied.)
+        bin_weights = 1.0 / np.where(bin_counts == 0, np.nan, bin_counts)
+
+        # For each sample in `truth_values`, get the weight in the corresponding bin
+        ix = np.digitize(truth[variable], bins=bins) - 1
+        sample_weights = bin_weights[ix]
+        sample_weights = sample_weights / sample_weights.mean()
+        if weight_name is None:
+            weight_name = variable + "_uniform_weight"
+        truth[weight_name] = sample_weights
+        truth = truth.drop(columns=[variable])
+        if add_to_database:
+            self._create_table(self._database_path, weight_name, truth)
+            save_to_sql(truth, weight_name, self._database_path)
+        return truth.sort_values("event_no").reset_index(drop=True)
+
+    def _get_truth(self, variable, selection):
+        if selection is None:
+            query = f"select {self._index_column}, {variable} from {self._truth_table}"
+        else:
+            query = f"select {self._index_column}, {variable} from {self._truth_table} where {self._index_column} in {str(tuple(selection))}"
+        with sqlite3.connect(self._database_path) as con:
+            data = pd.read_sql(query, con)
+        return data
+
+    def _create_table(self, database, table_name, df):
+        """Creates a table.
+        Args:
+            pipeline_database (str): path to the pipeline database
+            df (str): pandas.DataFrame of combined predictions
+        """
+        query_columns = list()
+        for column in df.columns:
+            if column == "event_no":
+                type_ = "INTEGER PRIMARY KEY NOT NULL"
+            else:
+                type_ = "FLOAT"
+            query_columns.append(f"{column} {type_}")
+        query_columns = ", ".join(query_columns)
+
+        code = (
+            "PRAGMA foreign_keys=off;\n"
+            f"CREATE TABLE {table_name} ({query_columns});\n"
+            "PRAGMA foreign_keys=on;"
+        )
+        run_sql_code(database, code)
+        return
```

Comment:
This seems to be duplicate of similar methods in [`graphnet.data.pipeline`](https://github.com/graphnet-team/graphnet/blob/main/src/graphnet/data/pipeline.py#L208-L229) and [`graphnet.data.sqlite.sqlite_dataconverter`](https://github.com/graphnet-team/graphnet/blob/main/src/graphnet/data/sqlite/sqlite_dataconverter.py#L120-L152). I think it would be nice to promote this to a utility function.